### PR TITLE
Root FS Builds with Ubuntu 20.04

### LINKED ISSUE #15: rootfs build fails
When building the rootfs in WSL 2 and in Docker on Arch Linux, the build fails. #14 makes it so the chroot works, but after getting into the chroot, package installation fails, saying there are unmet dependencies:
```
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 build-essential : Depends: libc6-dev but it is not going to be installed or
                            libc-dev
                   Depends: g++ (>= 4:7.2) but it is not going to be installed
                   Depends: dpkg-dev (>= 1.17.11) but it is not going to be installed
 network-manager : Depends: udev but it is not going to be installed
                   Depends: libpam-systemd but it is not going to be installed
                   Depends: policykit-1 but it is not going to be installed
 ssh : Depends: openssh-server (>= 1:7.6p1-4ubuntu0.3) but it is not going to be installed
```

I have never encountered this before, so I'm not sure what the root cause is. My guess is that it has something to do with including debian sources in our `sources.list`. I think those sources are there to install firmware for our Ralink wifi dongle. This was not an issue before updating the rootfs to 20.04, so maybe that is part of the cause. Either way, I don't think we should be including sources from debian in addition to sources from Ubuntu, as that seems like a good way to cause conflicts. 

Here's the full log:
```
archive already exists; using that one...

installing prerequisities...

Reading package lists... Done
Building dependency tree
Reading state information... Done
qemu-user-static is already the newest version (1:4.2-3ubuntu6.6).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

copying files to new rootfs...

updating packages...

Get:2 http://ports.ubuntu.com/ubuntu-ports bionic InRelease [242 kB]
Ign:1 http://cdn-fastly.deb.debian.org/debian jessie InRelease
Get:5 http://ports.ubuntu.com/ubuntu-ports bionic-updates InRelease [88.7 kB]
Ign:3 http://cdn-fastly.deb.debian.org/debian stretch InRelease
Get:4 http://cdn-fastly.deb.debian.org/debian jessie Release [77.3 kB]
Get:7 http://ports.ubuntu.com/ubuntu-ports bionic-backports InRelease [74.6 kB]
Get:8 http://ports.ubuntu.com/ubuntu-ports bionic-security InRelease [88.7 kB]
Get:6 http://cdn-fastly.deb.debian.org/debian stretch Release [118 kB]
Get:9 http://cdn-fastly.deb.debian.org/debian jessie Release.gpg [1652 B]
Get:10 http://cdn-fastly.deb.debian.org/debian stretch Release.gpg [2410 B]
Get:11 http://ports.ubuntu.com/ubuntu-ports bionic/main armhf Packages [968 kB]
Get:12 http://ports.ubuntu.com/ubuntu-ports bionic/main Translation-en [516 kB]
Get:13 http://ports.ubuntu.com/ubuntu-ports bionic/restricted armhf Packages [8360 B]
Get:14 http://ports.ubuntu.com/ubuntu-ports bionic/restricted Translation-en [3584 B]
Get:15 http://ports.ubuntu.com/ubuntu-ports bionic/universe armhf Packages [8269 kB]
Get:16 http://ports.ubuntu.com/ubuntu-ports bionic/universe Translation-en [4941 kB]
Get:17 http://ports.ubuntu.com/ubuntu-ports bionic/multiverse armhf Packages [129 kB]
Get:18 http://ports.ubuntu.com/ubuntu-ports bionic/multiverse Translation-en [108 kB]
Get:19 http://ports.ubuntu.com/ubuntu-ports bionic-updates/main armhf Packages [991 kB]
Get:20 http://ports.ubuntu.com/ubuntu-ports bionic-updates/main Translation-en [360 kB]
Get:21 http://ports.ubuntu.com/ubuntu-ports bionic-updates/restricted armhf Packages [10.0 kB]
Get:22 http://ports.ubuntu.com/ubuntu-ports bionic-updates/restricted Translation-en [23.0 kB]
Get:23 http://ports.ubuntu.com/ubuntu-ports bionic-updates/universe armhf Packages [1380 kB]
Get:24 http://ports.ubuntu.com/ubuntu-ports bionic-updates/universe Translation-en [352 kB]
Get:25 http://ports.ubuntu.com/ubuntu-ports bionic-updates/multiverse armhf Packages [6336 B]
Get:26 http://ports.ubuntu.com/ubuntu-ports bionic-updates/multiverse Translation-en [6920 B]
Get:27 http://ports.ubuntu.com/ubuntu-ports bionic-backports/main armhf Packages [9992 B]
Get:28 http://ports.ubuntu.com/ubuntu-ports bionic-backports/main Translation-en [4764 B]
Get:29 http://ports.ubuntu.com/ubuntu-ports bionic-backports/universe armhf Packages [9920 B]
Get:30 http://ports.ubuntu.com/ubuntu-ports bionic-backports/universe Translation-en [4588 B]
Get:31 http://ports.ubuntu.com/ubuntu-ports bionic-security/main armhf Packages [706 kB]
Get:32 http://ports.ubuntu.com/ubuntu-ports bionic-security/main Translation-en [269 kB]
Get:33 http://ports.ubuntu.com/ubuntu-ports bionic-security/restricted armhf Packages [5788 B]
Get:34 http://ports.ubuntu.com/ubuntu-ports bionic-security/restricted Translation-en [20.4 kB]
Get:35 http://ports.ubuntu.com/ubuntu-ports bionic-security/universe armhf Packages [851 kB]
Get:36 http://ports.ubuntu.com/ubuntu-ports bionic-security/universe Translation-en [239 kB]
Get:37 http://ports.ubuntu.com/ubuntu-ports bionic-security/multiverse armhf Packages [2368 B]
Get:38 http://ports.ubuntu.com/ubuntu-ports bionic-security/multiverse Translation-en [2908 B]
Ign:9 http://cdn-fastly.deb.debian.org/debian jessie Release.gpg
Get:39 http://cdn-fastly.deb.debian.org/debian jessie/main armhf Packages [6675 kB]
Ign:10 http://cdn-fastly.deb.debian.org/debian stretch Release.gpg
Get:40 http://cdn-fastly.deb.debian.org/debian jessie/main Translation-en [4581 kB]
Get:41 http://cdn-fastly.deb.debian.org/debian jessie/contrib armhf Packages [37.8 kB]
Get:42 http://cdn-fastly.deb.debian.org/debian jessie/contrib Translation-en [38.3 kB]
Get:43 http://cdn-fastly.deb.debian.org/debian jessie/non-free armhf Packages [62.5 kB]
Get:44 http://cdn-fastly.deb.debian.org/debian jessie/non-free Translation-en [72.0 kB]
Get:45 http://cdn-fastly.deb.debian.org/debian stretch/main armhf Packages [6908 kB]
Get:46 http://cdn-fastly.deb.debian.org/debian stretch/main Translation-en [5377 kB]
Get:47 http://cdn-fastly.deb.debian.org/debian stretch/contrib armhf Packages [41.8 kB]
Get:48 http://cdn-fastly.deb.debian.org/debian stretch/contrib Translation-en [45.8 kB]
Get:49 http://cdn-fastly.deb.debian.org/debian stretch/non-free armhf Packages [59.8 kB]
Get:50 http://cdn-fastly.deb.debian.org/debian stretch/non-free Translation-en [80.2 kB]
Fetched 44.9 MB in 47s (962 kB/s)
Reading package lists... Done
Building dependency tree
Reading state information... Done
1 package can be upgraded. Run 'apt list --upgradable' to see it.
W: GPG error: http://cdn-fastly.deb.debian.org/debian jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY CBF8D6FD518E17E1
W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY 04EE7237B7D453EC NO_PUBKEY EF0F382A1A7B6500
Reading package lists... Done
Building dependency tree
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  libseccomp2
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 32.7 kB of archives.
After this operation, 6144 B disk space will be freed.
Get:1 http://ports.ubuntu.com/ubuntu-ports bionic-updates/main armhf libseccomp2 armhf 2.4.3-1ubuntu3.18.04.3 [32.7 kB]
Fetched 32.7 kB in 1s (46.4 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 4114 files and directories currently installed.)
Preparing to unpack .../libseccomp2_2.4.3-1ubuntu3.18.04.3_armhf.deb ...
Unpacking libseccomp2:armhf (2.4.3-1ubuntu3.18.04.3) over (2.4.3-1ubuntu1) ...
Setting up libseccomp2:armhf (2.4.3-1ubuntu3.18.04.3) ...
Processing triggers for libc-bin (2.31-0ubuntu9) ...

installing packages...

Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 build-essential : Depends: libc6-dev but it is not going to be installed or
                            libc-dev
                   Depends: g++ (>= 4:7.2) but it is not going to be installed
                   Depends: dpkg-dev (>= 1.17.11) but it is not going to be installed
 network-manager : Depends: udev but it is not going to be installed
                   Depends: libpam-systemd but it is not going to be installed
                   Depends: policykit-1 but it is not going to be installed
 ssh : Depends: openssh-server (>= 1:7.6p1-4ubuntu0.3) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
/usr/bin/cp: cannot create regular file '/etc/avahi/services': No such file or directory
/usr/bin/cp: cannot create regular file '/etc/avahi/': Not a directory
/usr/bin/cp: cannot create regular file '/etc/ssh/': Not a directory
Selecting previously unselected package frost-edge.
(Reading database ... 4114 files and directories currently installed.)
Preparing to unpack /frost-edge.deb ...
Unpacking frost-edge (20201009) ...
Setting up frost-edge (20201009) ...
Moving to /var/lib/frost and cloning repos
/var/lib/dpkg/info/frost-edge.postinst: line 6: git: command not found
/var/lib/dpkg/info/frost-edge.postinst: line 7: cd: /var/lib/frost/deployment_manager: No such file or directory
/var/lib/dpkg/info/frost-edge.postinst: line 8: npm: command not found
/var/lib/dpkg/info/frost-edge.postinst: line 10: systemctl: command not found
/var/lib/dpkg/info/frost-edge.postinst: line 11: systemctl: command not found
/var/lib/dpkg/info/frost-edge.postinst: line 12: systemctl: command not found
dpkg: error processing package frost-edge (--install):
 installed frost-edge package post-installation script subprocess returned error exit status 127
Errors were encountered while processing:
 frost-edge
./setup_rootfs.sh: line 29: curl: command not found

./setup_rootfs.sh: line 34: curl: command not found

setting up user account...

done setting up rootfs...

Archiving ubuntu-base rootfs as rootfs.tar.gz...
```

### LINKED ISSUE #22: Updated .sh End Of Line Characters with Git Attributes
Updated the `.gitattributes` file:
```
* text=auto  eol=lf
```

This will eliminate the need for dos2unix conversions.

### PR DESCRIPTION
**Pull request by dack-fe**
Updated Root FS to Ubuntu 20.04 - Closes #15
Automatically adds Frost Edge and the supporting packages

Closes #22

### CODE REVIEW

**Comment by tvannoy** [comment]
The frost-edge deb does not get downloaded when the rootfs build is not done in Jenkins. Other than that, the docker build works for me on WSL. 

A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:
```
Err:1 http://ports.ubuntu.com/ubuntu-ports focal InRelease
  At least one invalid signature was encountered.
```
which causes a bunch of cascading failures when attempting to install packages. 

When installing the dotnet runtime, we get this warning: 
```
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.
```
Not a big deal, but we could look into using the installer instead of the script. When I set it up the first time, I found the dotnet script easier to use. 


Looks good overall

There are a couple of gpg key errors for some debian apt sources:
```
W: GPG error: http://cdn-fastly.deb.debian.org/debian jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY CBF8D6FD518E17E1
W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY 04EE7237B7D453EC NO_PUBKEY EF0F382A1A7B6500
```
I don't know if those sources are required. As best as I could tell, the `cdn-fastly` sources are commented out in our sources list. It's possible that we may not need the debian repos for the ralink firmware, but that's not very important as it seems to work as is. 

Looks good overall

File: rootfs/docker_build.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+docker build -t build_rootfs .
+docker run --name build_rootfs --privileged build_rootfs 
+#docker run --name build_rootfs --privileged --entrypoint /bin/bash -it build_rootfs
```
Comment (tvannoy) [inline rootfs/docker_build.sh]: delete commented out line?

File: rootfs/setup_rootfs.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
+# TODO: apt keeps complaining about our locale not being set; we should probably set it properly
```
Comment (tvannoy) [inline rootfs/setup_rootfs.sh]: I believe this is outdated now. It looks like locale/timezone are set up properly now.

**Comment by tvannoy** [comment]
We should be able to eliminate the need for using dos2unix by creating the following `.gitattributes` file:
```
* text=auto  eol=lf
```
We can use finer-grained control if we want, e.g. only sh files are forced to use lf endings in the repo and in checkout. 

https://www.git-scm.com/docs/gitattributes
https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

**Review by tvannoy — CHANGES_REQUESTED** [review]
We should add a short readme about how to execute the bulid process, e.g. run docker_build.sh

**Review by fe-tdavis — APPROVED** [review]

File: rootfs/setup_rootfs.sh (hunk 2 above)
Comment (dack-fe) [reply rootfs/setup_rootfs.sh]: 7a216c9

File: rootfs/docker_build.sh (hunk 1 above)
Comment (dack-fe) [reply rootfs/docker_build.sh]: 02d2ff7

**Comment by dack-fe** [comment]
> The frost-edge deb does not get downloaded when the rootfs build is not done in Jenkins. Other than that, the docker build works for me on WSL.
   - Updated with commit: c38b421 - Now the Build RootFS Script checks for Frost Edge from Jenkins and if it doesn't find it, the script downloads Frost Edge from Frost Release S3 Bucket.
---
> A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:
> ```
> Err:1 http://ports.ubuntu.com/ubuntu-ports focal InRelease
>   At least one invalid signature was encountered.
> ```
> 
> which causes a bunch of cascading failures when attempting to install packages.
   - This was never designed for a straight WSL Build due to permissions.
---
> 
> When installing the dotnet runtime, we get this warning:
> 
> ```
> dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
> dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
> dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
> dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.
> ```
> 
> Not a big deal, but we could look into using the installer instead of the script. When I set it up the first time, I found the dotnet script easier to use.
   - Added as Enhancement Issue #20  
---
> There are a couple of gpg key errors for some debian apt sources:
> 
> ```
> W: GPG error: http://cdn-fastly.deb.debian.org/debian jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY CBF8D6FD518E17E1
> W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY 04EE7237B7D453EC NO_PUBKEY EF0F382A1A7B6500
> ```
> 
> I don't know if those sources are required. As best as I could tell, the `cdn-fastly` sources are commented out in our sources list. It's possible that we may not need the debian repos for the ralink firmware, but that's not very important as it seems to work as is.
   - Added as Enhancement Issue #21

**Comment by dack-fe** [comment]
> We should be able to eliminate the need for using dos2unix by creating the following `.gitattributes` file:
> 
> ```
> * text=auto  eol=lf
> ```
> 
> We can use finer-grained control if we want, e.g. only sh files are forced to use lf endings in the repo and in checkout.
> 
> https://www.git-scm.com/docs/gitattributes
> https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

Added as Enhancement Issue #22

**Comment by tvannoy** [comment]
> This was never designed for a straight WSL Build due to permissions.
Gotcha. I think I had it working on WSL at some point, but I don't remember. I guess I would have likely needed to build it in the WSL filesystem. 

I didn't have a proper Ubuntu machine to test on, but I imagine it still works fine since that's what I originally developed on. Either way, docker handles any OS so it doesn't matter.

File: rootfs/ReadMe.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+## Key Files
+ - `build_rootfs.sh`
+ - `setup_rootfs.sh`
+ - `packages`
+
+## Building the Root FS with Docker
+1. Install [Docker for Windows](https://docs.docker.com/docker-for-windows/install/)
+2. Open WSL and Navigate to `[FrOST Repos]\utils\rootfs\`
+3. Run `./build_rootfs.sh`
```
Comment (tvannoy) [inline rootfs/ReadMe.md]: I (and the jenkinfiles) run `docker_build.sh`

Comment (dack-fe) [reply rootfs/ReadMe.md]: Updated with 2b9b109

**Comment by dack-fe** [comment]
> > This was never designed for a straight WSL Build due to permissions.
> > Gotcha. I think I had it working on WSL at some point, but I don't remember. I guess I would have likely needed to build it in the WSL filesystem.
> 
> I didn't have a proper Ubuntu machine to test on, but I imagine it still works fine since that's what I originally developed on. Either way, docker handles any OS so it doesn't matter.

Jenkins is currently building on Ubuntu 20.04 with Docker

**Review by tvannoy — APPROVED** [review]
👍 💯

**Comment by fe-wickham** [comment]
> A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:

Did you try running it directly through WSL on the Windows filesystem or the WSL filesystem? It will fail running it directly on the Windows filesystem because it can't handle the chroot properly. I recall getting errors in that same area but I don't know if they were the exact same error from running it on the Windows filesystem.

Edit:
>  I guess I would have likely needed to build it in the WSL filesystem.

Just saw you posted this, that is almost certainly the issue.

**Comment by fe-wickham** [comment]
> We should be able to eliminate the need for using dos2unix by creating the following `.gitattributes` file:
> 
> ```
> * text=auto  eol=lf
> ```
> 
> We can use finer-grained control if we want, e.g. only sh files are forced to use lf endings in the repo and in checkout.
> 
> https://www.git-scm.com/docs/gitattributes
> https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

I agree and needing to run `dos2unix` becomes rather annoying pretty quickly since things dont just work. I have fixed this with de1b2e9 which closes #22

**Review by fe-wickham — APPROVED** [review]
Looks good

**Comment by tvannoy** [comment]
> 
> 
> > A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:
> 
> Did you try running it directly through WSL on the Windows filesystem or the WSL filesystem? It will fail running it directly on the Windows filesystem because it can't handle the chroot properly. I recall getting errors in that same area but I don't know if they were the exact same error from running it on the Windows filesystem.
> 
> Edit:
> 
> > I guess I would have likely needed to build it in the WSL filesystem.
> 
> Just saw you posted this, that is almost certainly the issue.

Just built it in the WSL filesystem -- worked flawlessly. I now recall @fe-wickham mentioning chroot issues in the Windows filesystem before... 

I can also confirm that the frost-edge deb download worked, and there was no need to run dos2unix on a fresh clone, so the eol conversions seemed to work as well. 

It's beautiful! 🥳

**Comment by tvannoy** [comment]
I did get an npm error about certificates, though -- not sure if it's actually important:
```
npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
npm ERR! request to https://registry.npmjs.org/assert-plus/-/assert-plus-1.0.0.tgz failed, reason: unable to get local issuer certificate
```

I also get some warnings around sudo when the wifi-connect installer runs: 
```
sudo: unable to resolve host trevvvyx1carbon: Name or service not known
```
I suspect this is because we are running as root, so sudo doesn't really make sense. This isn't really an issue, but we could squelch that warning by using sed to remove `sudo` from the wifi-connect install script.

**Comment by dack-fe** [comment]
> I did get an npm error about certificates, though -- not sure if it's actually important:
> 
> ```
> npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> npm ERR! request to https://registry.npmjs.org/assert-plus/-/assert-plus-1.0.0.tgz failed, reason: unable to get local issuer certificate
> ```
 - I am not seeing these NPM Errors in Jenkins. @tvannoy Did you test the Frost Web App?

---
> I also get some warnings around sudo when the wifi-connect installer runs:
> 
> ```
> sudo: unable to resolve host trevvvyx1carbon: Name or service not known
> ```
> 
> I suspect this is because we are running as root, so sudo doesn't really make sense. This isn't really an issue, but we could squelch that warning by using sed to remove `sudo` from the wifi-connect install script.
 - We do not use a sudo call when configuring the WiFi

**Comment by tvannoy** [comment]
> > I did get an npm error about certificates, though -- not sure if it's actually important:
> > ```
> > npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> > npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> > npm ERR! request to https://registry.npmjs.org/assert-plus/-/assert-plus-1.0.0.tgz failed, reason: unable to get local issuer certificate
> > ```
> 
>  I am not seeing these NPM Errors in Jenkins. 

Interesting, I am getting them in the docker build and the WSL build. It appears to be solved by `npm config set strict-ssl false`, which is in frost-edge packaging scripts; I wonder if the version that is being downloaded from S3 does not have that npm setting. 

> @tvannoy Did you test the Frost Web App?

I am out of town, so I do not currently have an audio mini with me to test with.

> > I also get some warnings around sudo when the wifi-connect installer runs:
>
> We do not use a sudo call when configuring the WiFi

We don't, but [Balena's install script](https://raw.githubusercontent.com/balena-io/wifi-connect/master/scripts/raspbian-install.sh) that we download does.

**Comment by dack-fe** [comment]
Updated the Frost Edge Package on S3.

**Pull request merged**